cmd/kueue/main.go

```diff
@@ -278,7 +279,9 @@ func setupProbeEndpoints(mgr ctrl.Manager) {
 		setupLog.Error(err, "unable to set up health check")
 		os.Exit(1)
 	}
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+	if err := mgr.AddReadyzCheck("readyz", func(req *http.Request) error {
+		return mgr.GetWebhookServer().StartedChecker()(req)
+	}); err != nil {
```

Comment:
```suggestion
	if err := mgr.AddReadyzCheck("readyz",mgr.GetWebhookServer().StartedChecker()); err != nil {
```

Reply:
or could the `GetWebhookServer()` return different servers on different calls?

Reply:
I cannot do this, then the replicas never get ready. The function needs to get the webhook server once ready. I think the webhook server is set asynchronously in controller-runtime.

Reply:
.... very strange, maybe the cert manager is is doing it ... 
maybe add a comment about this if confirmed.

all other references in sigs are using it directly. 
https://github.com/search?q=org%3Akubernetes-sigs%20StartedChecker&type=code

Reply:
> , I guess we could do the same so that we don't get the transient errors logged.

What kind of errors? 

`GetWebhookServer().StartedChecker` should be the way to go since it actually attempts to connect the server.

Reply:
> I guess we could do the same so that we don't get the transient errors logged. WDYT?

I agree with you. We probably could use this:

https://github.com/kubernetes-sigs/kueue/blob/0b0961db512b5ee91047eaba6ff88694e46ff8c3/pkg/util/cert/cert.go#L70

> GetWebhookServer().StartedChecker should be the way to go since it actually attempts to connect the server.

IIUC, The `GetWebhookServer().StartedChecker` returns 200 only if the checker gets a positive response with tls from the webhook servers. So, we need to wait to be ready for webhook servers, which means we need to wait to be ready for certs.

In the case of `healthz.Ping` (our current checker), it always returns `true` since it doesn't check whether or not the webhook servers can actually work fine.

- `StartedChecker` : https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/webhook/server.go#L271-L273
- `healthz.Ping()`: https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/healthz/healthz.go#L186-L187

Reply:
I was thinking about the transient errors before the webhooks is ready, but they are very nice anyway:
```
2024-01-31T13:09:29.832263236Z stderr F 2024-01-31T13:09:29.832081013Z	DEBUG	controller-runtime.healthz	healthz/healthz.go:60	healthz check failed	{"checker": "readyz", "error": "webhook server has not been started yet"}
```
Maybe it doesn't actually make sense though to wait in the checker for ready certs, because it does not attempt to connect if the started=false: https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/webhook/server.go#L281C9-L281C16

And it changes started=true only after getting the certs: https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/webhook/server.go#L261 (just before listening)

Reply:
GetWebhookServer probably doesn't return a meaningful server until the first webhooks is registered, which only happens after the certs are ready.

Just leave a comment for it.

Reply:
> Maybe it doesn't actually make sense to wait in the checker for ready certs, because it does not attempt to connect if the started=false

I think that waiting for certs to get ready makes sense since the  `StartedChecker` returns a meaningful server only after the certs are ready:  https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/webhook/server.go#L237-L240

Reply:
`GetWebhookServer().StartedChecker` is what we need to determine the readiness.

I'm only not sure why we need a closure for calling it , since from what I can see the field  holding the webhook is private  and only stable whit an option when crating the manager,  however it's not a huge problem.

and

>>, I guess we could do the same so that we don't get the transient errors logged.

>What kind of errors?

Reply:
> `GetWebhookServer().StartedChecker` is what we need to determine the readiness.
> 
> I'm only not sure why we need a closure for calling it , since from what I can see the field holding the webhook is private and only stable whit an option when crating the manager, however it's not a huge problem.
> 

TBH, I'm not sure why either. What I verified experimentally is that if I just call `GetWebhookServer before `WaitForCertsReady` then the webhook is not initialized, and it stays that way. 

I get the following logs then even if keeping the readyz probe like currently, but just invoking `GetWebookServer()` before registering the probe:

```
2024-01-31T19:49:50.466048439Z	INFO	setup	kueue/main.go:291	readyz	{"ws": {"Options":{"Host":"","Port":9443,"CertDir":"/tmp/k8s-webhook-server/serving-certs","CertName":"tls.crt","KeyName":"tls.key","ClientCAName":"","TLSOpts":null,"WebhookMux":{}}}, "err": "webhook server has not been started yet"}
```

I suspect this has something to do with the one-time memorization of the not initialized webhook as runnable here: https://github.com/kubernetes-sigs/controller-runtime/blob/73519a939e27c8465c6c49694356cbd66daf1677/pkg/manager/internal.go#L258.

However, I'm not sure of the details. I would need to further investigate probably with debugger. Let me know if you have some ideas how to get to the bottom of it.

Let me put in the comment what I know so far.

> and
> 
> > > , I guess we could do the same so that we don't get the transient errors logged.
> 
> > What kind of errors?

I meant transient errors before the webhook server is listening, but I didn't how the look like in practice before writing. I was thinking maybe we have many errors about certs not ready. For what I can see we get these errors: 
```
2024-01-31T13:09:29.832263236Z stderr F 2024-01-31T13:09:29.832081013Z	DEBUG	controller-runtime.healthz	healthz/healthz.go:60	healthz check failed	{"checker": "readyz", "error": "webhook server has not been started yet"}
```
before the checker checks the `started` field before attempting the connection. So, there is only a small window where we can get transient errors, between setting `started: true`, and actually listening: https://github.com/kubernetes-sigs/controller-runtime/blob/8475c55f3e00e5611de0880eccd785efa85e8e38/pkg/webhook/server.go#L261-L263. However, I have never seen this errors so far.

Reply:
> I have never seen this errors so far.

I'ts normal behavior when the pod is not yet ready, (previously we ware "ready" from the start). 

Maybe something like: 
```suggestion
	if err := mgr.AddReadyzCheck("readyz", func(req *http.Request) error {
		select{
		case <-certsReady:
			return mgr.GetWebhookServer().StartedChecker()(req)
		case <-req.Context().Done():
			return errors.New("request canceled")
		}
	}); err != nil {
```
Could decrease the umber of messages.

But i find it a bit odd to block the http handler.

Reply:
I was thinking about waiting explicitly for `certsReady` (it is done in the `hierarchical-namespaces` project as pointed out above). 

However, I think it doesn't improve messages., because we already wait for certs ready as `mgr.GetWebhookServer().StartedChecker()` first checks if `started=true`. then, `started=true` means that certs are ready. If certs aren't ready we get the nice message: `{"checker": "readyz", "error": "webhook server has not been started yet"}`. I think this is enough, and no reason to complicate the code.

Reply:
Ok, I think I have some new insights into :

> I'm only not sure why we need a closure for calling it , since from what I can see the field holding the webhook is private and only stable whit an option when crating the manager, however it's not a huge problem.

if we call `GetWebhookServer`, then the not fully initialized webhook server runnable is registered here: https://github.com/kubernetes-sigs/controller-runtime/blob/73519a939e27c8465c6c49694356cbd66daf1677/pkg/manager/internal.go#L257. 

When manager starts it starts all runnables, including the webhook server, but it fails, in this case I see in logs early on:
```
2024-02-01T12:44:13.308211223Z	INFO	controller-runtime.webhook	webhook/server.go:191	Starting webhook server
2024-02-01T12:44:13.40926745Z	INFO	controller/controller.go:178	Starting EventSource	{"controller": "cert-rotator", "source": "kind source: *v1.Secret"}
2024-02-01T12:44:13.40932954Z	INFO	controller/controller.go:178	Starting EventSource	{"controller": "cert-rotator", "source": "kind source: *unstructured.Unstructured"}
2024-02-01T12:44:13.40933832Z	INFO	manager/internal.go:516	Stopping and waiting for non leader election runnables
```

Note the presence of "Stopping and waiting for non leader election runnables" just after the start.

OTOH (as in this PR) If we delay calling `GetWebhookServer` then the runnables are added after the certs are ready, and only then the webhook server is started, then it starts successfully. We the log `Starting webhook server` is much later (after certs ready).

Reply:
slightly polished the comment